### PR TITLE
Add rate limiting to authentication & registration endpoints.

### PR DESCRIPTION
**Pull request by DFurnes**
#### What's this PR do?
This pull request adds rate limiting to the authentication & registration endpoints: you can make 10 login attempts or registrations before hitting the cap, and 10 "invalid credentials" errors (so you can request as many tokens as you want, but are stopped if you keep making bad requests).

The limit resets after 15 minutes.

screen shot 2017-05-08 at 12 26 21 pm

#### How should this be reviewed?
This uses a slightly spruced up version of Laravel's rate limiting middleware (which changes the default cap & decay, and adds support for rate limiting browser requests). OAuth requests are rate-limited separately because we do not want to prevent internal clients from requesting a ton of tokens, so we specifically increment the limiter only on failed credentials.

Definitely let me know if there are other things we should be rate limiting that I missed. My goal here was to knock out the most "dangerous" things while not introducing any unexpected bugs for internal clients that rely on Northstar and don't expect to be slowed down. 🐌

This PR is a bit hefty, but should be okay if you go commit-by-commit. Tests in d446766.

#### Checklist
- [ ] Documentation added for changed endpoints.
- [ ] Tests added for new features/bug fixes.
- [ ] Post a message in #api if this includes something that causes a rebuild!

### CODE REVIEW

**Comment by DFurnes** [comment]
This is behind a `DS_ENABLE_RATE_LIMITING` feature flag so we can disable it on QA until [it can report accurate IP addresses](https://github.com/DoSomething/devops/issues/161#issuecomment-299277284), or in case we run into unexpected issues when enabling on production. ⚠️

File: app/Listeners/RateLimitedRequest.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    /**
+     * Handle the event.
+     *
+     * @return void
+     */
+    public function handle()
+    {
+        $this->stathat->ezCount('rate-limited user authentication attempt');
```
Comment (deadlybutter) [inline app/Listeners/RateLimitedRequest.php]: just a random note- id be curious to see what this looks like after a few days in the wild.

File: app/Providers/EventServiceProvider.php [hunk 2]
```diff
@@ -32,6 +34,14 @@ public function boot(DispatcherContract $events)
     {
         parent::boot($events);
 
+        // Rate limit failed client authentication attempts.
+        // @see: OAuthController::createToken
+        $oauth = app(AuthorizationServer::class);
+        $oauth->getEmitter()->addListener('client.authentication.failed', function () {
+            // Increment number of failed requests for this route & IP address.
+            app(RateLimiter::class)->hit(request()->fingerprint(), 1);
```
Comment (deadlybutter) [inline app/Providers/EventServiceProvider.php]: I'm guessing this library handles the reset as well?

File: app/Http/Controllers/OAuthController.php [hunk 3]
```diff
@@ -72,10 +83,24 @@ public function info()
      *
      * @param ServerRequestInterface $request
      * @param ResponseInterface $response
-     * @return \Illuminate\Http\Response
+     * @return ResponseInterface
+     * @throws OAuthServerException
      */
     public function createToken(ServerRequestInterface $request, ResponseInterface $response)
     {
+        $shouldRateLimit = config('features.rate-limiting');
```
Comment (weerd) [inline app/Http/Controllers/OAuthController.php]: nice :)

File: app/Http/Controllers/Web/AuthController.php [hunk 4]
```diff
@@ -191,4 +194,14 @@ public function postRegister(Request $request)
 
         return redirect()->intended('/');
     }
+
+    /**
+     * Get the login username to be used by the controller.
+     *
+     * @return string
+     */
+    public function loginUsername()
```
Comment (weerd) [inline app/Http/Controllers/Web/AuthController.php]: Just curious, why did you separate this?

File: tests/Http/OAuthTest.php [hunk 5]
```diff
@@ -187,6 +187,44 @@ public function testClientCredentialsGrantWithInvalidCredentials()
...
+
+    /**
+     * Test that requests are rate limited after 10 invalid attempts.
+     */
+    public function testInvalidClientCredentialsAreRateLimited()
+    {
+        $invalidCredentials = [
+            'grant_type' => 'client_credentials',
+            'client_id' => 'phpunit',
+            'client_secret' => 'banana',
```
Comment (weerd) [inline tests/Http/OAuthTest.php]: 🍌

File: app/Http/Controllers/Web/AuthController.php (hunk 4 above)
Comment (DFurnes) [inline app/Http/Controllers/Web/AuthController.php]: Ah, I should probably change that back. I'd played around with the `ThrottlesLogins` built-in trait before realizing that didn't really work well for our other throttling needs and it needs that to exist (based on how the default authorization controller trait).

File: app/Providers/EventServiceProvider.php (hunk 2 above)
Comment (DFurnes) [inline app/Providers/EventServiceProvider.php]: Yep! That's all handled in the `tooManyAttempts` method on the limiter, and is a rolling 15 minute window.

File: app/Listeners/RateLimitedRequest.php (hunk 1 above)
Comment (DFurnes) [inline app/Listeners/RateLimitedRequest.php]: Agreed! Hopefully not too high. 🔒

File: .env.example [hunk 6]
```diff
@@ -13,9 +13,9 @@ DS_ENABLE_PASSWORD_RESET=true
 # DB_AUTH_NAME=admin
 # DB_SSL=false
 
-# If running behind a load balancer, specify which
-# addresses to trust `X-Forwarded-For` headers from.
-# TRUSTED_PROXY_IP_ADDRESS=
+# If running behind a load balancer, specify which IP addresses (for
+# example, HAProxy or Fastly) to trust `X-Forwarded-For` headers from.
+# TRUSTED_PROXY_IP_ADDRESSES=
```
Comment (DFurnes) [inline .env.example]: Ah, good idea!

**Comment by DFurnes** [comment]
@weerd @deadlybutter Added a few more commits based on feedback!

**Comment by weerd** [comment]
🐑 it!

**Comment by DFurnes** [comment]
[🔪 As a security-conscious organization, I want to rate limit things like user registration. 🔪](https://trello.com/c/gBglBwYO/411-as-a-security-conscious-organization-i-want-to-rate-limit-things-like-user-registration)

**Pull request merged**